cpp/src/io/parquet/experimental/hybrid_scan_impl.cpp

```diff
@@ -0,0 +1,274 @@
+/*
+ * Copyright (c) 2025, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "hybrid_scan_impl.hpp"
+
+#include "hybrid_scan_helpers.hpp"
+
+// #include "error.hpp"
+#include <cudf/detail/stream_compaction.hpp>
+#include <cudf/detail/transform.hpp>
+#include <cudf/detail/utilities/stream_pool.hpp>
+#include <cudf/strings/detail/utilities.hpp>
+#include <cudf/utilities/error.hpp>
+#include <cudf/utilities/memory_resource.hpp>
+
+#include <thrust/iterator/counting_iterator.h>
+
+#include <bitset>
+#include <iterator>
+#include <limits>
+#include <numeric>
+
+namespace cudf::experimental::io::parquet::detail {
+
+impl::impl(cudf::host_span<uint8_t const> footer_bytes,
+           cudf::host_span<uint8_t const> page_index_bytes,
+           cudf::io::parquet_reader_options const& options)
+{
+  // Open and parse the source dataset metadata
+  _metadata = std::make_unique<aggregate_reader_metadata>(
+    footer_bytes,
+    page_index_bytes,
+    options.is_enabled_use_arrow_schema(),
+    options.get_columns().has_value() and options.is_enabled_allow_mismatched_pq_schemas());
+
+  // Strings may be returned as either string or categorical columns
+  auto const strings_to_categorical = options.is_enabled_convert_strings_to_categories();
+
+  // Select only columns required by the filter
+  std::optional<std::vector<std::string>> filter_columns_names;
+  if (options.get_filter().has_value() and options.get_columns().has_value()) {
+    // list, struct, dictionary are not supported by AST filter yet.
+    // extract columns not present in get_columns() & keep count to remove at end.
+    filter_columns_names = cudf::io::parquet::detail::get_column_names_in_expression(
+      options.get_filter(), *(options.get_columns()));
+    _num_filter_only_columns = filter_columns_names->size();
+  }
+
+  // Only need to select columns if filter is available
+  if (options.get_filter().has_value()) {
```

Comment:
Not my favorite piece of code to get `filter_column_names` and `select` them but temporarily putting here.